packages/exo/src/exo-tools.js

```diff
@@ -6,59 +6,150 @@ import {
+ * enforcement for a method guard, and is implied by all other
+ * non-raw method guards.
+ */
+const PassableMethodGuard = M.call().rest(M.any()).returns(M.any());
+
+/**
+ * @typedef {object} RedactConfig
+ * @property {number} declaredLen
+ * @property {boolean} hasRestArgGuard
+ * @property {Pattern} paramsPattern
+ * @property {boolean} rawRestArgs
+ * @property {number[]} redactedIndices
  */
-const MinMethodGuard = M.call().rest(M.any()).returns(M.any());
 
 /**
  * @param {Passable[]} syncArgs
- * @param {MethodGuardPayload} methodGuardPayload
+ * @param {RedactConfig} redactConfig
  * @param {string} [label]
  * @returns {Passable[]} Returns the args that should be passed to the
  * raw method
  */
-const defendSyncArgs = (syncArgs, methodGuardPayload, label = undefined) => {
-  const { argGuards, optionalArgGuards, restArgGuard } = methodGuardPayload;
-  const paramsPattern = M.splitArray(
-    argGuards,
-    optionalArgGuards,
-    restArgGuard,
-  );
-  mustMatch(harden(syncArgs), paramsPattern, label);
-  if (restArgGuard !== undefined) {
+const defendSyncArgs = (syncArgs, redactConfig, label = undefined) => {
+  const {
+    declaredLen,
+    hasRestArgGuard,
+    paramsPattern,
+    rawRestArgs,
+    redactedIndices,
+  } = redactConfig;
+
+  // Only manipulate the original args if our params pattern didn't contain any
+  // raw values.
+  let redactedArgs = syncArgs;
+  if (rawRestArgs) {
+    // Don't harden the rest args.
+    redactedArgs = syncArgs.slice(0, Math.min(syncArgs.length, declaredLen));
+  } else if (redactedIndices.length > 0) {
+    // Copy the arguments array, avoiding hardening the redacted ones (which are
+    // trivially matched).
+    redactedArgs = [...syncArgs];
+  }
+
+  for (const i of redactedIndices) {
+    if (i >= redactedArgs.length) {
+      break;
+    }
+    redactedArgs[i] = REDACTED_RAW_ARG;
+  }
+
+  mustMatch(harden(redactedArgs), paramsPattern, label);
+
+  if (hasRestArgGuard) {
     return syncArgs;
   }
-  const declaredLen =
-    argGuards.length + (optionalArgGuards ? optionalArgGuards.length : 0);
   if (syncArgs.length <= declaredLen) {
     return syncArgs;
   }
   // Ignore extraneous arguments, as a JS function call would do.
   return syncArgs.slice(0, declaredLen);
 };
 
+/**
+ * @param {MethodGuardPayload} methodGuardPayload
+ */
+const redactRawArgs = methodGuardPayload => {
+  const {
+    argGuards,
+    optionalArgGuards = [],
+    restArgGuard,
+  } = methodGuardPayload;
+
+  const rawArgGuards = [...argGuards, ...optionalArgGuards];
+
+  const redactedIndices = [];
+  for (let i = 0; i < rawArgGuards.length; i += 1) {
+    if (isRawGuard(rawArgGuards[i])) {
+      rawArgGuards[i] = REDACTED_RAW_ARG;
+      redactedIndices.push(i);
+    }
+  }
+
+  // Always match the redacted args if there is a raw rest that applies.
+  let rawRestArgGuard = restArgGuard;
+  let rawRestArgs = false;
+  if (isRawGuard(rawRestArgGuard)) {
+    rawRestArgGuard = undefined;
```

Comment:
```suggestion
    rawRestArgGuard = REDACTED_RAW_ARG;
```

Reply:
This needed some finessing, after replacing `undefined` with the only pattern that could possibly work: `M.arrayOf(REDACTED_RAW_ARG)`.

Reply:
If we're going through the effort to create rest arguments that are each `REDACTED_RAW_ARG`, I'd rather do the stricter `arrayOf` pattern.  Without being strict, I see no compelling reason to do anything other than just drop raw rest arguments rather than doing this whole match-but-bypass dance.